rust/src/parse.rs

```diff
@@ -0,0 +1,392 @@
+            Self::StringEncoding(err) => write!(f, "string field not valid utf-8: {}", err),
+        }
+    }
+}
+
+impl Error for ParseError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            ParseError::StringEncoding(err) => Some(err),
+            _ => None,
+        }
+    }
+}
+
+impl From<std::str::Utf8Error> for ParseError {
+    fn from(err: std::str::Utf8Error) -> Self {
+        ParseError::StringEncoding(err)
+    }
+}
+
+fn parse_u16(data: &[u8]) -> Result<(u16, &[u8]), ParseError> {
+    if data.len() < std::mem::size_of::<u16>() {
+        return Err(ParseError::DataTooShort);
+    }
+    let (int_bytes, data) = data.split_at(std::mem::size_of::<u16>());
+    Ok((
+        u16::from_le_bytes(int_bytes.try_into().expect("expected 2 bytes")),
+        data,
+    ))
+}
+
+fn parse_u32(data: &[u8]) -> Result<(u32, &[u8]), ParseError> {
+    if data.len() < std::mem::size_of::<u32>() {
+        return Err(ParseError::DataTooShort);
+    }
+    let (int_bytes, data) = data.split_at(std::mem::size_of::<u32>());
+    Ok((
+        u32::from_le_bytes(int_bytes.try_into().expect("expected 4 bytes")),
+        data,
+    ))
+}
+
+pub(crate) fn parse_u64(data: &[u8]) -> Result<(u64, &[u8]), ParseError> {
+    if data.len() < std::mem::size_of::<u64>() {
+        return Err(ParseError::DataTooShort);
+    }
+    let (int_bytes, data) = data.split_at(std::mem::size_of::<u64>());
+    Ok((
+        u64::from_le_bytes(int_bytes.try_into().expect("expected 8 bytes")),
+        data,
+    ))
+}
+
+pub(crate) fn parse_byte_array(data: &[u8]) -> Result<(&[u8], &[u8]), ParseError> {
+    let (len, data) = parse_u32(data)?;
+    Ok(data.split_at(len as usize))
+}
+
+pub(crate) fn parse_long_byte_array(data: &[u8]) -> Result<(&[u8], &[u8]), ParseError> {
+    let (len, data) = parse_u64(data)?;
+    Ok(data.split_at(len as usize))
+}
+
+fn parse_str(data: &[u8]) -> Result<(&str, &[u8]), ParseError> {
+    let (str_len, data) = parse_u32(data)?;
+    if data.len() < str_len as usize {
+        return Err(ParseError::DataTooShort);
+    }
+    let (str_bytes, data) = data.split_at(str_len as usize);
+    Ok((std::str::from_utf8(str_bytes)?, data))
+}
+
+type StrMap<'a> = BTreeMap<Cow<'a, str>, Cow<'a, str>>;
+
+fn parse_str_map(data: &[u8]) -> Result<(StrMap<'_>, &[u8]), ParseError> {
+    let (map_len, data) = parse_u32(data)?;
+    let (map_data, remainder) = data.split_at(map_len as usize);
+    let mut result: BTreeMap<Cow<'_, str>, Cow<'_, str>> = BTreeMap::new();
+    let mut unparsed_map_data = map_data;
+    {
+        while !unparsed_map_data.is_empty() {
+            let (key, data) = parse_str(unparsed_map_data)?;
+            let (val, data) = parse_str(data)?;
+            unparsed_map_data = data;
+            result.insert(key.into(), val.into());
+        }
+    }
+    Ok((result, remainder))
+}
+
+fn parse_header(data: &[u8]) -> Result<Header, ParseError> {
```

Comment:
any hope of reducing boilerplate by doing these with a macro?

Would it make sense to make them static methods on the record types themselves, e.g. something like `Header::try_from(data: &[u8])`? That's what I did in the Swift library

Reply:
TryFrom isn't a bad idea, not sure why I didn't do that.